core/src/main/java/com/linecorp/armeria/server/annotation/BeanRequestConverterFunction.java

```diff
@@ -0,0 +1,546 @@
+/*
+ * Copyright 2018 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.server.annotation;
+
+import static com.linecorp.armeria.internal.AnnotatedHttpServiceParamUtil.convertParameter;
+import static com.linecorp.armeria.internal.AnnotatedHttpServiceParamUtil.httpParametersOf;
+import static com.linecorp.armeria.internal.AnnotatedHttpServiceParamUtil.validateAndNormalizeSupportedType;
+import static org.reflections.ReflectionUtils.getAllFields;
+import static org.reflections.ReflectionUtils.getAllMethods;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.AnnotatedElement;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Executable;
+import java.lang.reflect.Field;
+import java.lang.reflect.Member;
+import java.lang.reflect.Method;
+import java.lang.reflect.Parameter;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentMap;
+import java.util.stream.Collectors;
+
+import org.reflections.ReflectionUtils;
+
+import com.google.common.collect.MapMaker;
+
+import com.linecorp.armeria.common.AggregatedHttpMessage;
+import com.linecorp.armeria.common.HttpParameters;
+import com.linecorp.armeria.internal.AnnotatedHttpServiceParamUtil.EnumConverter;
+import com.linecorp.armeria.server.ServiceRequestContext;
+
+import io.netty.util.AsciiString;
+
+/**
+ * A default {@link RequestConverterFunction} which converts a request into a Java bean.
+ *
+ * <p>The following parameters are eligible for automatic injection into bean fields or properties:
+ * <ul>
+ *   <li>Path parameters to fields annotated by {@link Param}</li>
+ *   <li>HTTP parameters to fields annotated by {@link Param}</li>
+ *   <li>Headers to fields annotated by {@link Header}</li>
+ * </ul>
+ */
+public class BeanRequestConverterFunction implements RequestConverterFunction {
+    private static final String PACKAGE_NAME_PREFIX_JAVA = "java.";
+
+    private static final ConcurrentMap<Class<?>, BeanParamInfo> beanParamInfoCache = new MapMaker().weakKeys()
+                                                                                                   .makeMap();
```

Comment:
After thinking a little bit about this, I realized that the entries in this map will never be garbage-collected because `BeamParamInfo` will have strong references to `Field`, `Method` and `Constructor`, which end up with strong references to its parent `Class`.

I think it's not worth fix this properly by making all reflective fields `WeakReference`s given their performance and complexity problem. How about just using `ConcurrentHashMap` for `beamParamInfoCache` and adding some comment about this?

Reply:
fixed